google-cloud-datastore/lib/google/cloud/datastore/dataset.rb

```diff
@@ -509,8 +518,26 @@ def run query, namespace: nil, consistency: nil
         #     tx.rollback
         #   end
         #
-        def transaction
-          tx = Transaction.new service
+        # @example Use a read-only transaction when only performing reads:
+        #   require "google/cloud/datastore"
+        #
+        #   datastore = Google::Cloud::Datastore.new
+        #
+        #   task_list_key = datastore.key "TaskList", "default"
+        #   query = datastore.query("Task").
+        #     ancestor(task_list_key)
+        #
+        #   tasks = nil
+        #
+        #   datastore.transaction read_only: true do |tx|
+        #     task_list = tx.find task_list_key
+        #     if task_list
+        #       tasks = tx.run query
+        #     end
+        #   end
+        #
+        def transaction read_only: nil
```

Comment:
I would strongly prefer that the `read_only` option uses a separate named `snapshot` method, similar to what is done in Spanner's `Client#snapshot`.

Reply:
I'm afraid that users will not be expecting to create a read-only `Transaction` using a method named `snapshot`, and will struggle to understand the relationship. The official docs do not mention "Snapshot" as a first-class concept.* And we have no `Snapshot` class.

Is consistency with spanner worth it?

* The official docs state: "you can use a transaction to read a consistent snapshot of Cloud Datastore"

Reply:
I think there are many advantages to supporting a read-write transaction in the Transaction class, and supporting a read-only transaction in a Snapshot class. One is that you can omit the `delete`, `insert`, `save`, and `update` methods completely.

Spanner's Snapshot class is a read-only transaction, and is a concept specific to the client library. Meaning, there is no Snapshot resource in the Spanner API. I don't think doing the same in Datastore would cause undue confusion.

In general, I think Snapshot is a useful abstraction and one that I would like to be consistent across Datastore, Spanner, and Firestore.

Reply:
Chiming into the discussion:
As a user having a restricted object that doesn't include write operations would be helpful so I don't incorrectly use a read-only Transaction as a read-write transaction. 

Using a Snapshot class seems like it would be useful considering the work on the Spanner read-only transactions are created using a snapshot method. Datastore [documentation](https://cloud.google.com/datastore/docs/concepts/transactions) does use the word snapshot even though it's not necessarily part of the [API endpoints or parameters](https://cloud.google.com/datastore/docs/reference/rest/v1/projects/beginTransaction#ReadOnly) similar to the Spanner [API reference documentation](https://cloud.google.com/spanner/docs/reference/rest/v1/TransactionOptions#ReadOnly.FIELDS.strong), but Spanner reference documentation documents ReadOnly transactions as "snapshots".

I'm on the fence of using "snapshot" because it's not very self descriptive such as `datastore.transaction read_only: true` or maybe even `datastore.read_only_transaction`. In [Php PR#727](https://github.com/GoogleCloudPlatform/google-cloud-php/pull/727/files#diff-2124f3325a11419ecc0faf5cd6e8c598R524) may use a similar approach as the `datastore.read_only_transaction` method.

I agree there should exist an object that restricts the methods available to a read-only transaction versus a read-write transaction. But I'm not so sure about using Snapshot as the surfaced name.